src/core/lib/promise/party.cc

```diff
@@ -246,136 +220,55 @@ void Party::RunLocked() {
-  uint64_t prev_state;
-  do {
-    // Grab the current state, and clear the wakeup bits & add flag.
-    prev_state = state_.fetch_and(kRefMask | kLocked | kAllocatedMask,
-                                  std::memory_order_acquire);
-    if (grpc_trace_promise_primitives.enabled()) {
-      gpr_log(GPR_DEBUG, "%s[party] Run prev_state=%s", DebugTag().c_str(),
-              StateToString(prev_state).c_str());
-    }
-    GPR_ASSERT(prev_state & kLocked);
-    if (prev_state & kDestroying) return true;
-    // From the previous state, extract which participants we're to wakeup.
-    uint64_t wakeups = prev_state & kWakeupMask;
-    // Now update prev_state to be what we want the CAS to see below.
-    prev_state &= kRefMask | kLocked | kAllocatedMask;
-    // For each wakeup bit...
-    for (size_t i = 0; wakeups != 0; i++, wakeups >>= 1) {
-      // If the bit is not set, skip.
-      if ((wakeups & 1) == 0) continue;
-      // If the participant is null, skip.
-      // This allows participants to complete whilst wakers still exist
-      // somewhere.
-      auto* participant = participants_[i].load(std::memory_order_acquire);
-      if (participant == nullptr) {
-        if (grpc_trace_promise_primitives.enabled()) {
-          gpr_log(GPR_DEBUG, "%s[party] wakeup %" PRIdPTR " already complete",
-                  DebugTag().c_str(), i);
-        }
-        continue;
-      }
-      absl::string_view name;
+  return sync_.RunParty([this](int i) {
+    // If the participant is null, skip.
+    // This allows participants to complete whilst wakers still exist
+    // somewhere.
+    auto* participant = participants_[i].load(std::memory_order_acquire);
+    if (participant == nullptr) {
       if (grpc_trace_promise_primitives.enabled()) {
-        name = participant->name();
-        gpr_log(GPR_DEBUG, "%s[%s] begin job %" PRIdPTR, DebugTag().c_str(),
-                std::string(name).c_str(), i);
+        gpr_log(GPR_DEBUG, "%s[party] wakeup %d already complete",
+                DebugTag().c_str(), i);
       }
-      // Poll the participant.
-      currently_polling_ = i;
-      if (participant->Poll()) {
-        if (!name.empty()) {
-          gpr_log(GPR_DEBUG, "%s[%s] end poll and finish job %" PRIdPTR,
-                  DebugTag().c_str(), std::string(name).c_str(), i);
-        }
-        participants_[i] = nullptr;
-        const uint64_t allocated_bit = (1u << i << kAllocatedShift);
-        prev_state &= ~allocated_bit;
-        state_.fetch_and(~allocated_bit, std::memory_order_release);
-      } else if (!name.empty()) {
-        gpr_log(GPR_DEBUG, "%s[%s] end poll", DebugTag().c_str(),
-                std::string(name).c_str());
+      return false;
+    }
+    absl::string_view name;
+    if (grpc_trace_promise_primitives.enabled()) {
+      name = participant->name();
+      gpr_log(GPR_DEBUG, "%s[%s] begin job %d", DebugTag().c_str(),
+              std::string(name).c_str(), i);
```

Comment:
nit: can we use `name.data()` or is this because you have to have it null terminated?

Reply:
it's got to be null terminated... this is printf

(we'll get absl logging "soon")